WebAssembly toolchain

diff --git a/clang/lib/Driver/ToolChains/WebAssembly.cpp b/clang/lib/Driver/ToolChains/WebAssembly.cpp
index 36bed3166..3ecd055df 100644
--- a/clang/lib/Driver/ToolChains/WebAssembly.cpp
+++ b/clang/lib/Driver/ToolChains/WebAssembly.cpp
@@ -51,7 +51,10 @@ std::string wasm::Linker::getLinkerPath(const ArgList &Args) const {
     }
   }
 
-  return ToolChain.GetProgramPath(ToolChain.getDefaultLinker());
+  const char *defaultLinker = ToolChain.getDefaultLinker();
+  if (llvm::sys::path::is_absolute(defaultLinker))
+    return defaultLinker;
+  return ToolChain.GetProgramPath(defaultLinker);
 }
 
 void wasm::Linker::ConstructJob(Compilation &C, const JobAction &JA,
@@ -123,8 +126,9 @@ void wasm::Linker::ConstructJob(Compilation &C, const JobAction &JA,
     }
 
     CmdArgs.push_back("-lc");
-    AddRunTimeLibs(ToolChain, ToolChain.getDriver(), CmdArgs, Args);
   }
+  if (!Args.hasArg(options::OPT_nodefaultlibs))
+    AddRunTimeLibs(ToolChain, ToolChain.getDriver(), CmdArgs, Args);
 
   CmdArgs.push_back("-o");
   CmdArgs.push_back(Output.getFilename());
